### PR TITLE
Added support for debuggability

### PR DESCRIPTION
**Pull request by MISBMS**
Added support for debuggability, with a default implementation.  It serves as a reference for custom implementations in future. This for now, doesn't factor in pre-operations support (in the value that would show up in the detail). It may be provided in future.

### CODE REVIEW

File: src/main/java/io/appform/jsonrules/Expression.java [hunk 1]
```diff
@@ -74,8 +75,7 @@
         @JsonSubTypes.Type(name = "in", value = InExpression.class),
         @JsonSubTypes.Type(name = "not_in", value = NotInExpression.class),
         @JsonSubTypes.Type(name = "contains_any", value = ContainsAnyExpression.class),
-        @JsonSubTypes.Type(name = "contains_all", value = ContainsAllExpression.class),
-})
+        @JsonSubTypes.Type(name = "contains_all", value = ContainsAllExpression.class), })
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/Expression.java:78]: Revert

File: src/main/java/io/appform/jsonrules/Expression.java [hunk 2]
```diff
@@ -48,8 +50,7 @@
 @EqualsAndHashCode
 @ToString
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "type")
-@JsonSubTypes({
-        @JsonSubTypes.Type(name = "equals", value = EqualsExpression.class),
+@JsonSubTypes({ @JsonSubTypes.Type(name = "equals", value = EqualsExpression.class),
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/Expression.java:53]: Revert

File: src/main/java/io/appform/jsonrules/expressions/debug/DenialDetail.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import io.appform.jsonrules.ExpressionType;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+@AllArgsConstructor
+public class DenialDetail {
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/debug/DenialDetail.java:30]: Maybe rename to `FailureDetail`

File: src/main/java/io/appform/jsonrules/expressions/debug/ExpressionDebugger.java [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+                .reason(details.stream()
+                        .map(DenialDetail::getReason)
+                        .flatMap(Collection::stream)
+                        .collect(Collectors.toList()))
+                .build();
+    }
+
+    @Override
+    public DenialDetail visit(OrExpression expression, JsonNode node) {
+        final List<DenialDetail> details = expression.getChildren()
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/debug/ExpressionDebugger.java:91]: Save for one line, the rest of the code is same with previous method. Refactor.

File: src/main/java/io/appform/jsonrules/expressions/debug/ExpressionDebugger.java [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+        final boolean result = expression.evaluate(node);
+        if (result) {
+            return DEFAULT_SUCCESS_RESPONSE;
+        }
+        return DenialDetail.builder()
+                .expressionType(expression.getType())
+                .path(expression.getPath())
+                .value(value)
+                .denied(true)
+                .reason(Collections.singletonList(String.format("Value of [%s] at path [%s] is not between [%s] & [%s]",
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/expressions/debug/ExpressionDebugger.java:447]: All of these can be refactored by passing the formatted string i guess?

File: src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java [hunk 6]
```diff
@@ -16,111 +16,113 @@
...
 import java.time.OffsetDateTime;
 import java.time.ZoneOffset;
 import java.time.temporal.ChronoField;
 
+import com.fasterxml.jackson.databind.JsonNode;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java:30]: Replace with `@UtilityClass` annotation

File: src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java [hunk 7]
```diff
@@ -16,111 +16,113 @@
...
+            return dateTime.get(ChronoField.DAY_OF_YEAR);
+        case WEEK_OF_MONTH:
+            return dateTime.get(ChronoField.ALIGNED_WEEK_OF_MONTH);
+        case WEEK_OF_YEAR:
+            return dateTime.get(ChronoField.ALIGNED_WEEK_OF_YEAR);
+        case MONTH_OF_YEAR:
+            return dateTime.get(ChronoField.MONTH_OF_YEAR);
+        default:
+        }
+        throw new IllegalArgumentException("Operand doesnot represent a valid field");
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java:61]: does not

File: src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java [hunk 8]
```diff
@@ -16,111 +16,113 @@
...
-		} catch (NumberFormatException exception) {
-			return false;
-		}
-		return true;
-	}
+    private static boolean isEpochInMillis(long epoch) {
+        // Safe check to know if epoch is in milli seconds for all dates between 1973 to
+        // 5138
+        return String.valueOf(epoch)
+                .length() >= 12;
```
Comment (santanusinha) [inline src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java:125]: This check is problematic. Something like `Instant.ofEpochMilli(epoch)` is better. Refer to [this](https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html#ofEpochMilli-long-)

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java (hunk 8 above)
Comment (MISBMS) [reply src/main/java/io/appform/jsonrules/utils/PreOperationUtils.java:124]: This is to figure out if the epoch given is in millis or seconds. And accordingly, an `Instant` is created.

**Review by santanusinha — APPROVED** [review]

**Review by santanusinha — APPROVED** [review]

**Pull request merged**